### DISCUSSION TITLE
Import/Export in Core

### DISCUSSION THREAD

**Discussion by angrybrad** [craftcms/cms#9179] [Ideas]
> Created by: Florian (fw@demodern.de) on 2015/03/13 06:38:29 +0000
> Votes at time of UserVoice import: 6

-----------------------------------

Actually believed this would have been one of the first Feature requests and was surprised to not see it here.

We have a ticket for "syncing" db's between installs which is basically an extension of a simple export/import between craft installs.

So here we go with this feature request: Clean Exports out of craft (and decide what can be exported and clean imports data/structure into craft)

**Comment by timkelty** [comment]
Throwing in my 2 cents here:

This would be crucial for us. On practically every project, we stumble through importing data, sinking tons of time and frustration into it. 

Yes, there are plugins to do this. In my experience they have always been buggy, error-prone, with limited features. They can also wreak havoc on your database in ways you may not even be immediately obvious.

**Why should this be included in core vs plugin?**

- A big reason is testing and reproducibly. In a system as dynamic as a CMS, content is often the culprit of uncaught bugs. Being able to consistently import the same data and run tests would a big step.
- As CMSs go, this is a pretty common, and frequently "expected" feature.

**Comment by khalwat** [comment]
yes please

A core function of a CMS should be to make it easy to get my data in/out

**Comment by shaunagm** [comment]
I'm using Craft because someone else selected it for a project, and this has been the biggest pain point for us.  I've considered redoing the entire project to move to a new CMS that has solid import/export but haven't had the time yet.  I don't like being so critical in an issue tracker, because I know how hard it is to prioritize the thousand features users always want and I'm sure you're doing your best, but you should know there's at least one person actively warning people away from Craft because of this.

**Comment by timkelty** [comment]
Well said, @shaunagm. Good to hear a perspective from someone not already entrenched in the Craft ecosystem.

**Comment by narration-sd** [comment]
For the most, I should keep out of this, as have opinions and not the huge weight of need you who enable large sites do, as @angrybrad and @brandonkelly both know.

What I can offer though is why I'm sitting here working late nights and Saturday, which relates.

People in our closed Beta, when they finally opened up, just _had_ to have portable, repo-able, scriptable content for the much smaller database of Live Vue - so they can assure the presence of Craft Live Previews for content editors on their headless Spas.

It was a considerable push, even if we're just doing Gatsby first, and I heard the feelings. I'm making it -- in yaml for community lingua franca, and I'm handling all they don't realize of intricate issues inside, making them go away. 

I know it's going to be really worth it, arriving soon. 
I took the pause on other features and opportunities to get here.

**Comment by brandonkelly** [comment]
Just added basic element exporting to all element index pages, for Craft 3.2:

<img width="327" alt="Screen Shot 2019-04-02 at 7 33 52 PM" src="https://user-images.githubusercontent.com/47792/55449280-bce76680-557f-11e9-8db4-46d19e98ce54.png">

To help test, change your `craftcms/cms` requirement in composer.json to:

```json
"require": {
  "craftcms/cms": "3.2.x-dev as 3.2.0-alpha.1",
  "...": "..."
}
```

Then run `composer update`.

**Comment by narration-sd** [comment]
Well, that's pretty amazing work, @brandonkelly Brandon.

You got me to try 3.2 as well :) . And, bang, _both_ items Just Worked :tm: .  20 minutes when I am well fatigued, winding down with graphics after succeeding on the last Parkour among the mazes.

It's on a server I could afford to knock down, and there was only one Entry there. But it came through immediately with what looks like its database surely.

You can see the Spanish song lyric from this Excel view -- I tried csv and xslx. Great contempo use of components, btw, And I will have to show you the robot that's begun to inhabit Live Vue...it's more than charm.

Respect.

Clive

![brandon-s-surprise](https://user-images.githubusercontent.com/247945/55452077-c32f1000-558a-11e9-83b7-8d2f66db80b4.png)

**Comment by narration-sd** [comment]
Oh, yes. For anyone trying this and getting hung up with composer because they have PHP on their dev machine, but not extension to run phpoffice/spreadsheet, I learned a new thing this evening:

```
composer update  --ignore-platform-reqs
```

... will get you there.

**Comment by shaunagm** [comment]
Thank you, @brandonkelly!

**Comment by brandonkelly** [comment]
We just announced that [we’ve acquired Feed Me](https://craftcms.com/blog/feed-me), a popular content import plugin for Craft, and we’ve made it free.

It’s a first step toward implementing content importing directly into core :)

**Comment by narration-sd** [comment]
Great stuff, @brandonkelly - congratulations!

**Comment by ryanpcmcquen** [comment]
Does anyone have a current solution for migrating data across multiple environments? Full database migration is not viable for us.

**Comment by cosly** [comment]
Why is the Matrix field type left out of it?

**Comment by brandonkelly** [comment]
@cosly All custom fields are currently excluded from export. We will be working to improve export options and also introduce element importing in a future release, which is why this issue is still open.

**Comment by carlcs** [comment]
https://github.com/craftcms/cms/issues/4484
@cosly see this related issue

**Comment by KatieMFritz** [comment]
What's a good workaround for exporting entries with custom fields while we wait for Craft 4? Is there a recommended plugin for this? 

https://github.com/superbigco/craft-reports seems unfinished, and https://github.com/kffein/Craft-export-Csv doesn't support relations or nested fields either. :/

**Comment by brandonkelly** [comment]
You could use [Element API](https://plugins.craftcms.com/element-api) to create a custom JSON feed that has exactly what you need in it.

**Comment by sjelfull** [comment]
@KatieMFritz For what its worth, Reports is unfinished but working. Feel to reach out directly with any questions or feedback.

**Comment by daltonrooney** [comment]
@KatieMFritz GraphQL is a credible export solution as well. I was able to get it up and running and learn the syntax very quickly and now I’m thinking of building my whole front-end with it! But it would also be a good way to quickly create a JSON file with all of your entry data including custom fields, matrix fields, etc.

**Comment by foxtrotcz** [comment]
I was just looking how to easily import/export entries in Craft.
Export with Element Exporter is really easy but importing with Feed Me requires me to map all fields manually...

For example, I have local dev site and production site. All structures between them are the same because I can just copy the project config. But I cant easily copy entries... (other than dump DB and do it manually...).

So I hope for some solution that lets me easily import entries exported with element exporter.
Thanks.

**Reply by mini71** [reply]
It would be extremely helpful to have such an easy import feature for entries. 
We often prepare pages with content changes on a test/preview system. Moving them to production always ends up in doing the editor work again.
Feed Me isn't designed for this.

@brandonkelly Is something on your roadmap?

**Reply by brandonkelly** [reply]
Yeah we want to make it possible to export the content + metadata that could be used to auto-map on import

**Reply by natebeaty** [reply]
Just wanted to toss in my 2¢ that I'm very often doing this, especially when we're nearing a launch with a staging site where the client is actively editing content, and we're still doing some tweaking and content testing in dev, so the full db export/import becomes impossible. The export option in core is so nice, it would be great to have even a basic variation of Feed Me to complement it.

**Comment by myleshyson** [comment]
Hey just wondering where this stands? I find myself wanting to import specific entries into our testing/local environments a lot. Right now I have to sync the entire production database with my local database over a vpn connection and it takes forever.

I could see this either being a feedme feature or a simple cli command, or both. I wouldn't mind proposing something if there isn't already something else in the works.

**Reply by timbertens** [reply]
I have good experiences with https://github.com/dgrigg/craft-migration-assistant
Until recently it was only supported till v3, but recently Derrick picked up development again and made the component compatible with v4, support for v5 is on its way. I'm not affiliated with migration assistant in any way but we use it intensively for multiple sites that have the similar content and it works pretty well. Content migration can be made from the control panel and create regular migration files that can automatically imported using existing CLI commands, something that the Verbb plugin is missing. Since it is feature unique in that sense I believe it would be a good candidate to become first party plugin if you ask me just like it was done for Feedme in the past

**Comment by daltonrooney** [comment]
@myleshyson Verbb’s [Zen](https://verbb.io/craft-plugins/zen/features) plugin does this and in my testing it works well. A little pricey to license it per-site, I wish there was a way to get a developer license for a plugin that will be used on multiple sites.

**Reply by myleshyson** [reply]
Oof yea that is a little pricey. It's definitely feature rich, but I feel like Craft is already halfway there in therms of core functionality. It's just missing an import UI. 

Feedme has the necessary components for this I think. It would just need an "auto map" option on the field mapping page that tries to map the export file properties to fields automatically. Definitely a little more involved than dragging and dropping a file though.

I think a CLI command could be a good intermediary step between no entry import and a new UI in Feedme. Thinking it can work like the `wp import` command.
